XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,419 @@
+then only the `destroy()` callback needs to be passed.
+
+**Error Handling**: If any callback throws the application will print the stack
+trace and exit. The exit path does follow that of any uncaught exception,
+except for the fact that it is not catchable by an uncaught exception handler,
+so any `'exit'` callbacks will fire. Unless the application is run with
+`--abort-on-uncaught-exception`. In which case a stack trace will be printed
+and the application will exit, leaving a core file.
+
+The reason for this behavior is that these callbacks are running at potentially
+volatile points in an object's lifetime. For example during class construction
+and destruction. Because of this, it is deemed necessary to bring down the
+process quickly as to prevent an unintentional abort in the future. This is
+subject to change in the future if a comprehensive analysis is performed to
+ensure an exception can follow the normal control flow without unintentional
+side effects.
+
+
+#### `asyncHook.scope()`
+
+Enable capture of asynchronous events until the current synchronous code
+execution has completed. This is basically a small amount of sugar to prevent
+accidentally forgetting to `disable()` a set of hooks in a function with
+multiple returns. Also in case an error is thrown and caught by a domain or
+`uncaughtException`.
+
+Say an application only wishes to observe the asynchronous calls made from
+another asynchronous callback. Without `scope()` the call would look like so:
+
+```js
+fs.readFile(path, (err, data) => {
+  asyncHook.enable();
+  var e = asyncDataProcessing(data);
+  if (e) {
+    asyncHook.disable();
+    return;
+  }
+  doMoreAsyncStuff();
+  asyncHook.disable();
+});
+```
+
+But using `scope()` all of that is handled for you:
+
+```js
+fs.readFile(path, (err, data) => {
+  asyncHook.scope();
+  var e = asyncDataProcessing(data);
+  if (e)
+    return;
+  doMoreAsyncStuff();
+});
+```
+
+
+#### `asyncHook.enable()`
+
+Enable the callbacks for a given `AsyncHook` instance. Once a callback fires on
+an asynchronous event they will continue to fire for all nested asynchronous
+events. Even after the instance has been disabled.
+
+Callbacks are not implicitly enabled after an instance is created. The reason
+for this is to not make any assumptions about the user's use case. Since
+constructing the `asyncHook` during startup, but not using it until later is, a
+perfectly reasonable use case. This API is meant to err on the side of
+requiring explicit instructions from the user.
+
+
+#### `asyncHook.disable()`
+
+Disable the callbacks for a given `AsyncHook` instance. Doing this will prevent
+the `init()`, etc., calls from firing for any new roots of asynchronous call
+stacks, but will not prevent existing asynchronous call stacks that have
+already been captured by the `AsyncHook` instance from continuing to fire.
+
+While not part of the immediate development plan, it should be possible in the
```

Comment:
This is all nice, but what does it do now? What would the code sample below do with the current development plan? Disable the hook in its entirety rather than for just the current call stack? It would be good to clarify below the example what the current behaviour is. Also, what does the migration path look like to making this functionality work if it doesn't now? Would it be a breaking change to the API to make it work as is being proposed here for future work as opposed to what goes in now?